lib/avromatic/model/attributes.rb

```diff
@@ -163,33 +163,25 @@ def define_avro_attributes(schema, allow_optional: true)
             attribute_definition = AttributeDefinition.new(
               owner: self,
               field: field,
-              type: create_type(field)
+              type: Avromatic::Model::Types::TypeFactory.create(schema: field.type, nested_models: nested_models)
             )
             attribute_definitions[symbolized_field_name] = attribute_definition
 
-            define_method(field.name) { _attributes[symbolized_field_name] }
-            define_method("#{field.name}?") { !!_attributes[symbolized_field_name] } if boolean?(field)
+            # Add all generated methods to a module so they can be overridden
+            generated_methods_module.send(:define_method, field.name) { _attributes[symbolized_field_name] }
+            generated_methods_module.send(:define_method, "#{field.name}?") { !!_attributes[symbolized_field_name] } if FieldHelper.boolean?(field)
 
-            define_method("#{field.name}=") do |value|
+            generated_methods_module.send(:define_method, "#{field.name}=") do |value|
               _attributes[symbolized_field_name] = attribute_definitions[symbolized_field_name].coerce(value)
             end
 
             unless config.mutable # rubocop:disable Style/Next
-              private("#{field.name}=")
-              define_method(:clone) { self }
-              define_method(:dup) { self }
+              generated_methods_module.send(:private, "#{field.name}=")
+              generated_methods_module.send(:define_method, :clone) { self }
+              generated_methods_module.send(:define_method, :dup) { self }
             end
           end
         end
-
-        def boolean?(field)
```

Comment:
I moved this to `FieldHelper` and inlined `create_type` to avoid polluting the model class' method namespace.